### PR TITLE
[Validator] Modernize Custom constraints article (finishes #13898)

### PR DESCRIPTION
**Pull request by wouterj**
Replaces #13898

While the main addition of that PR was unit tests (which we merged from a concurrent PR), it also brought some great other things: Real & consistent examples.
Let's get them in the docs :)

Also, I've added the type declarations to this article (like we are doing in most examples at the moment).

### CODE REVIEW

File: validation/custom_constraint.rst [hunk 1]
```diff
@@ -194,22 +196,62 @@ Class Constraint Validator
...
-    public function getTargets()
+    use Symfony\Component\Validator\Constraint;
+
+    /**
+     * @Annotation
+     */
+    class ConfirmedPaymentReceipt extends Constraint
     {
-        return self::CLASS_CONSTRAINT;
+        public string $userDoesNotMatchMessage = 'User\'s e-mailaddress does not match that of the receipt';
```
Comment (xabbuh) [inline validation/custom_constraint.rst]:
```suggestion
        public string $userDoesNotMatchMessage = 'User\'s e-mail address does not match that of the receipt';
```

File: validation/custom_constraint.rst [hunk 2]
```diff
@@ -194,22 +196,62 @@ Class Constraint Validator
...
+
+        public function getTargets(): string
+        {
+            return self::CLASS_CONSTRAINT;
+        }
     }
 
-With this, the validator's ``validate()`` method gets an object as its first argument::
+Now, the constraint validator will get an object as first argument to
```
Comment (xabbuh) [inline validation/custom_constraint.rst]:
```suggestion
Now, the constraint validator will get an object as the first argument to
```

File: validation/custom_constraint.rst [hunk 3]
```diff
@@ -221,67 +263,76 @@ With this, the validator's ``validate()`` method gets an object as its first arg
...
+        class PaymentReceipt
         {
             // ...
         }
 
     .. code-block:: yaml
 
         # config/validator/validation.yaml
-        App\Entity\AcmeEntity:
+        AppBundle\Entity\PaymentReceipt:
```
Comment (xabbuh) [inline validation/custom_constraint.rst]:
```suggestion
        App\Entity\PaymentReceipt:
```

**Review by xabbuh — APPROVED** [review]
with some minor suggestions

**Review by javiereguiluz — APPROVED** [review]

**Comment by javiereguiluz** [comment]
Thanks Wouter.

**Pull request merged**